internal/pipeline/executor.go

```diff
@@ -36,159 +38,129 @@ type Executor struct {
+	if err != nil {
+		klog.Errorf("failed to write pkg: %v", err)
+		return fmt.Errorf("failed to write pkg: %w", err)
+	}
 	return nil
 }
 
+// helper function to log debug information about a resource.
+func debugResource(r *yaml.RNode) {
+	meta, _ := r.GetMeta()
+	klog.Infof("resource %s annotations: %v", meta.Name, meta.Annotations)
+}
+
 //
 // hydrationContext contains bits to track state of a package hydration.
 // This is sort of global state that is available to hydration step at
 // each pkg along the hydration walk.
 type hydrationContext struct {
 	// root points to the root of hydration graph where we bagan our hydration journey
-	root *pkg
-
-	// TODO (droot): wire sink-mode here
+	root *pkgNode
 
-	// pkgs refers to the discovered pkgs where
-	// It's a map where key refers to package's absolute path and value
-	// refers to the pkg.
-	pkgs map[string]*pkg
-
-	// hydrationStates for each package (dry, hydrating or wet) in the hydration DAG
-	hydrationStates map[string]hydrationState
+	// pkgs refers to the packages undergoing hydration.
+	pkgs map[string]*pkgNode
 }
 
 //
-// pkg represents a kpt package (node) in the hydration DAG.
+// pkgNode represents a package being hydrated. Think of it as a node in the hydration DAG.
 //
-// TODO(droot): kpt will certainly have pkg abstraction, may be embed that here to
-// avoid duplication.
-type pkg struct {
-	// absolute path on the filesystem
-	path string
+type pkgNode struct {
+	pkg *pkg.Pkg
 
-	// pipeline contained in the pkg
-	pipeline *kptfilev1alpha2.Pipeline
+	// state indicates if the pkg is being hydrated or done.
+	state hydrationState
 
-	// resources post hydration
+	// KRM resources that we have gathered post hydration for this package.
+	// These inludes resources at this pkg as well all it's children.
 	resources []*yaml.RNode
 }
 
-// TODO(droot): this will be replaced when pkg abstraction PR is merged
-func newPkg(path string) (*pkg, error) {
-	pkgPath, err := filepath.Abs(path)
+// newPkgNode returns a pkgNode instance given a path.
+func newPkgNode(path string) (*pkgNode, error) {
+	p, err := pkg.New(path)
 	if err != nil {
-		return nil, fmt.Errorf("failed to find abs path %s %w", path, err)
+		return nil, fmt.Errorf("failed to read pkg %w", err)
 	}
-	p := &pkg{
-		path: pkgPath,
+
+	pn := &pkgNode{
+		pkg:   p,
+		state: Dry, // package starts in dry state
 	}
-	return p, nil
+	return pn, nil
 }
 
-func (p *pkg) Path() string {
-	return p.path
+func (pn *pkgNode) Path() string {
+	return string(pn.pkg.UniquePath)
 }
 
-func (p *pkg) Read() error {
-	// Read the pipeline for the current package else assume default
-	// TODO(droot): integrate with Donny's code once that is merged
-	// and read an actual pipeline
-	p.pipeline = &kptfilev1alpha2.Pipeline{}
-
-	return nil
-}
+// subPackages returns paths to sub packages including current package's path.
+// It includes a package only if it has a Kptfile.
+func (pn *pkgNode) subPackages() ([]string, error) {
```

Comment:
This functionality should be extracted out/abstracted and used across different features (I'm guessing this is re-implemented multiple times already):

pkg.New(path).SubPackages() -> []Pkg

I don't think there's any reason for hydration logic to do any low-level file operations.

Reply:
Good point. I moved this to `pkg` abstraction.